synapse/federation/federation_base.py

```diff
@@ -98,9 +98,14 @@ async def _check_sigs_and_hash(
                 )
             return redacted_event
 
-        result = await self.spam_checker.check_event_for_spam(pdu)
+        reject_hard = await self.spam_checker.reject_federated_spam_event(pdu)
+        if reject_hard:
+            logger.warning("Event contains spam, hard-failing %s", pdu.event_id)
+            raise FederationDeniedError(get_domain_from_id(pdu.sender))
```

Comment:
Sorry, as I was double checking the code paths that call this I think this is actually the wrong place to put it. Instead, can we move it to:

https://github.com/matrix-org/synapse/blob/0d17357fcdded3fa3f8a37db7b6b9aa0402a10ed/synapse/federation/federation_server.py#L1021

which gets called when we receive an event directly over federation and before it goes into the staging area,

In fact, maybe we should have it before *and* after we pull events from the staging area, so that if we change the rules while there are events in the staging area the rules get applied to them too. I think the appropriate place is just before we call `on_receive_pdiu`:

https://github.com/matrix-org/synapse/blob/0d17357fcdded3fa3f8a37db7b6b9aa0402a10ed/synapse/federation/federation_server.py#L1088